README.rst
===
bcc
===


.. image:: https://img.shields.io/pypi/v/bcc.svg
        :target: https://pypi.python.org/pypi/bcc

.. image:: https://img.shields.io/travis/willsheffler/bcc.svg
        :target: https://travis-ci.org/willsheffler/bcc

.. image:: https://readthedocs.org/projects/bcc/badge/?version=latest
        :target: https://bcc.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status




N-dim invertible BCC lattice


* Free software: Apache Software License 2.0
* Documentation: https://bcc.readthedocs.io.


Features
--------

* TODO

Credits
-------

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


HISTORY.rst
=======
History
=======

0.1.9 (2018-07-10)
------------------

* add numba support

0.1.8 (2018-03-12)
------------------

* Basic tests in place

0.1.7 (2018-03-06)
------------------

* First release on PyPI.


tests/test_bcc.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for `bcc` package."""

import pytest


from bcc import bcc


@pytest.fixture
def response():
    """Sample pytest fixture.

    See more at: http://doc.pytest.org/en/latest/fixture.html
    """
    # import requests
    # return requests.get('https://github.com/audreyr/cookiecutter-pypackage')


def test_content(response):
    """Sample pytest test function with the pytest fixture as an argument."""
    # from bs4 import BeautifulSoup
    # assert 'GitHub' in BeautifulSoup(response.content).title.string


bcc/tests/__init__.py
# -*- coding: utf-8 -*-

"""Unit test package for bcc."""


bcc/tests/test_bcc.py
"""Tests for `bcc` package."""

import pytest
from bcc import *
from hypothesis import given, settings
import hypothesis.strategies as hs

try:
    import numba
    only_if_numba = lambda f: f
except ImportError:
    import pytest
    only_if_numba = pytest.mark.skip

# try:
#     from rif.numeric.lattice import *
#     import rif
#     import rif.util
#     only_if_have_rif = lambda x: x
#     oracles = [None, None, None, BCC3, BCC4, BCC5,
#                BCC6, BCC7, BCC8, BCC9, BCC10]
# except ImportError:
#     only_if_have_rif = pytest.mark.skip

max_examples = 1


def len_or_m_one(x):
    try:
        return len(x)
    except TypeError:
        return -1


def point_grid(sizes, lb, ub):
    n = max(len_or_m_one(x) for x in (lb, ub, sizes))
    lb = [lb] * n if len_or_m_one(lb) is -1 else lb
    ub = [ub] * n if len_or_m_one(ub) is -1 else ub
    sizes = [sizes] * n if len_or_m_one(sizes) is -1 else sizes
    linsp = [np.linspace(lb[i], ub[i], sizes[i]) for i in range(len(lb))]
    return np.stack(np.meshgrid(*linsp), axis=-1).reshape(-1, len(lb))


def index_samples(mx, nsamp):
    if mx < nsamp:
        return np.arange(mx)
    else:
        return np.random.randint(0, mx, nsamp)


@hs.composite
def sizes_lb_ub(draw, min_ndim=3, max_ndim=10):
    ndim = draw(hs.integers(min_value=min_ndim, max_value=max_ndim))
    sizes = draw(
        hs.lists(
            hs.integers(min_value=2, max_value=10),
            min_size=ndim,
            max_size=ndim))
    lb0 = draw(
        hs.lists(
            hs.floats(
                min_value=-10,
                max_value=10,
                allow_nan=False,
                allow_infinity=False),
            min_size=ndim,
            max_size=ndim))
    ub0 = draw(
        hs.lists(
            hs.floats(
                min_value=-10,
                max_value=10,
                allow_nan=False,
                allow_infinity=False),
            min_size=ndim,
            max_size=ndim))
    return sizes, list(np.minimum(lb0, ub0)), list(np.maximum(ub0, ub0) + 1)


def test_BCC_traits():
    with pytest.raises(tl.TraitError):
        BCC().sizes
    with pytest.raises(tl.TraitError):
        BCC(sizes=[1, 2, 3], lower=[1, 2]).width
    with pytest.raises(tl.TraitError):
        BCC(sizes=[1, 2, 3], upper=[1, 2]).width
    with pytest.raises(tl.TraitError):
        BCC(sizes=[[1, 2], [3, 4]]).upper
    with pytest.raises(tl.TraitError):
        BCC(sizes=[1, 2], ).upper
    with pytest.raises(tl.TraitError):
        BCC(sizes=[1, 2], lower=[1, 1], upper=[0, 2])
    assert np.allclose(BCC(sizes=(3, 4, 5)).lower, [0, 0, 0])
    assert np.allclose(BCC(sizes=(3, 4, 5)).upper, [1, 1, 1])
    assert np.allclose(BCC(sizes=(2, 4, 8)).width, [0.5, 0.25, 0.125])


# result: python/numpy 2-3 times slower than cpp
# def test_BCC_against_rifcpp_perf():
#    import sys
#    # sys.path.append('/home/sheffler/rifsrc/buildR/lib.linux-x86_64-3.6')
#    sys.path.insert(0, '/home/sheffler/rifsrc/buildR/lib.linux-x86_64-3.6')
#    import rif.util
#    from rif.numeric.lattice import BCC6
#    from time import clock
#
#    sizes = [100] * 6
#    lb = [0] * 6
#    ub = [10] * 6
#
#    binr = BCC(sizes=sizes, lower=lb, upper=ub)
#    oracle = BCC6(sizes, lb, ub)
#
#    # print(i0[i0 != i1])
#    # print(i1[i0 != i1])
#
#    for nside_points in range(2, 16, 2):
#        linsp = [np.linspace(lb[i], ub[i], nside_points)
#                 for i in range(len(sizes))]
#        test_points = np.stack(np.meshgrid(*linsp), axis=-1)
#        t0 = clock()
#        i0 = oracle.index(test_points)
#        # c0 = oracle.center(test_indices)['raw']
#        t0, t1 = clock() - t0, clock()
#        i1 = binr.get_bin_index(test_points)
#        # c1 = binr.get_bin_center(test_indices)
#        t1 = clock() - t1
#        print('numpy vs rifcpp get_bin_index', nside_points, t1 / t0)
#
#    for i in (1, 10, 100, 1000, 10000, 100000, 1000000):
#        test_indices = index_samples(len(binr), i)
#        t0 = clock()
#        c0 = oracle.center(test_indices)['raw']
#        t0, t1 = clock() - t0, clock()
#        c1 = binr.get_bin_center(test_indices)
#        t1 = clock() - t1
#        print('numpy vs rifcpp get_bin_center', i, t1 / t0)
#
#    assert 0

# @only_if_have_rif
# def test_BCC_against_rifcpp_case0():
#     sizes = [2, 2, 2, 2,  # 3,
#              3, 3, 3]
#     lb = [0.0, 0.0, 0.0, 0.0,  # 0.0,
#           # 0.4687500000000000,  # works
#           # 0.4687500000000001,  # fails
#           0.46876,  # fails
#           # 0.468751,  # works
#           0.0,
#           0.012699127197266515]
#     ub = [1.0, 1.0, 1.0, 1.0,  # 1.0,
#           1.46875, 1.0, 7.987300872802734]

#     print(np.array(ub) - np.array(lb))
#     # assert 0

#     binr = BCC(sizes=sizes, lower=lb, upper=ub)
#     oracle = oracles[len(sizes)](sizes, lb, ub)

#     nside_points = (10000**(1 / len(sizes)))
#     linsp = [np.linspace(lb[i], ub[i], nside_points)
#              for i in range(len(sizes))]
#     test_points = np.stack(np.meshgrid(*linsp), axis=-1)
#     i0 = oracle.index(test_points)
#     i1 = binr.get_bin_index(test_points)
#     print(i0.shape)
#     print(sizes, lb, ub)
#     # print(i0[i0 != i1])
#     # print(i1[i0 != i1])
#     assert np.all(i0 == i1)

#     test_indices = index_samples(len(binr), 10000)
#     c0 = oracle.center(test_indices)['raw']
#     c1 = binr.get_bin_center(test_indices)
#     assert np.allclose(c0, c1)

# @only_if_have_rif
# @given(sizes_lb_ub())
# @settings(max_examples=max_examples)
# def test_BCC_against_rifcpp(inp):
# _test_BCC_against_rifcpp(inp)


@given(sizes_lb_ub())
@settings(max_examples=max_examples)
def test_BCC_invertibility_by_index(inp):
    sizes, lb, ub = inp
    binr = BCC(sizes=sizes, lower=lb, upper=ub)
    idx0 = index_samples(len(binr), 50000)
    cen = binr.get_bin_center(idx0)
    cen_inbounds = 0 == (
        np.sum(cen < binr.lower, axis=-1) + np.sum(cen > binr.upper, axis=-1))
    idx0 = idx0[cen_inbounds]
    cen = cen[cen_inbounds]
    idx1 = binr.get_bin_index(cen)
    all_idx0_idx1_same = np.all(idx0 == idx1)
    print('fraction wrong:', sum(idx0 != idx1) / len(idx0))
    print(cen[idx0 != idx1])
    print(idx0[idx0 != idx1])
    print(idx1[idx0 != idx1])
    assert all_idx0_idx1_same


@given(sizes_lb_ub())
@settings(max_examples=max_examples)
def test_BCC_invertibility_by_center(inp):
    sizes, lb, ub = inp
    binr = BCC(sizes=sizes, lower=lb, upper=ub)
    pts = (binr.lower +
           (np.random.rand(10000, len(sizes)) * (binr.upper - binr.lower)))
    assert np.all(binr.lower <= pts)
    assert np.all(binr.upper >= pts)
    idx0 = binr.get_bin_index(pts)
    assert np.all(idx0 >= 0)
    cen = binr.get_bin_center(idx0)
    idx1 = binr.get_bin_index(cen)
    print('fraction wrong:', sum(idx0 != idx1) / len(idx0))
    print(idx0[idx0 != idx1])
    print(idx1[idx0 != idx1])
    assert np.all(idx0 == idx1)


@pytest.mark.skip
@given(sizes_lb_ub(min_ndim=4, max_ndim=4))
@settings(max_examples=max_examples)  # , max_shrinks=1, timeout=36000)
def test_BCC_covrad(inp):
    sizes, lb, ub = (np.array(x) for x in inp)
    ndim = len(sizes)
    binr = BCC(sizes=sizes, lower=lb, upper=ub)

    pts = (
        binr.lower + (np.random.rand(50000, ndim) * (binr.upper - binr.lower)))
    assert np.all(binr.lower <= pts)
    assert np.all(binr.upper >= pts)
    idx = binr.get_bin_index(pts)
    assert np.all(idx >= 0)
    cen = binr.get_bin_center(idx)
    err = np.linalg.norm((pts - cen) / binr.width, axis=-1)

    # from testing, not math... should figure out?
    unit_cov_rad = [
        0.559,  # ndim == 3
        0.706,  # ndim == 4
        0.748,  # ndim == 5
        0.858,  # ndim == 6
        0.893,  # ndim == 7
        0.984,  # ndim == 8
        1.010,  # ndim == 9
        1.079,  # ndim == 10
    ]
    correct = np.all(err < unit_cov_rad[ndim - 3])
    if not correct:
        print('test_BCC_covrad_ndim', ndim, 'err', np.max(err))
    assert correct


@only_if_numba
def test_numba_get_bin_index():
    nside = np.array([10] * 6, dtype='i4')
    lower = np.array([0] * 6, dtype='f8')
    upper = np.array([1] * 6, dtype='f8')

    bcc = BCC(sizes=nside, lower=lower, upper=upper)
    indexer = numba_bcc_indexer(nside, lower, upper)
    guindexer = gu_bcc_indexer(nside, lower, upper)

    for i in range(100):
        r = np.random.random(6)
        idx0 = bcc.get_bin_index(r)
        idx1 = indexer(r)
        assert idx0 == idx1

    r = np.random.random((10000, 6))
    idx0 = bcc.get_bin_index(r)
    idx1 = guindexer(r)
    assert np.all(idx0 == idx1)


travis_pypi_setup.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Update encrypted deploy password in Travis config file."""


from __future__ import print_function
import base64
import json
import os
from getpass import getpass
import yaml
from cryptography.hazmat.primitives.serialization import load_pem_public_key
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.asymmetric.padding import PKCS1v15


try:
    from urllib import urlopen
except ImportError:
    from urllib.request import urlopen


GITHUB_REPO = 'willsheffler/bcc'
TRAVIS_CONFIG_FILE = os.path.join(
    os.path.dirname(os.path.abspath(__file__)), '.travis.yml')


def load_key(pubkey):
    """Load public RSA key.

    Work around keys with incorrect header/footer format.

    Read more about RSA encryption with cryptography:
    https://cryptography.io/latest/hazmat/primitives/asymmetric/rsa/
    """
    try:
        return load_pem_public_key(pubkey.encode(), default_backend())
    except ValueError:
        # workaround for https://github.com/travis-ci/travis-api/issues/196
        pubkey = pubkey.replace('BEGIN RSA', 'BEGIN').replace('END RSA', 'END')
        return load_pem_public_key(pubkey.encode(), default_backend())


def encrypt(pubkey, password):
    """Encrypt password using given RSA public key and encode it with base64.

    The encrypted password can only be decrypted by someone with the
    private key (in this case, only Travis).
    """
    key = load_key(pubkey)
    encrypted_password = key.encrypt(password, PKCS1v15())
    return base64.b64encode(encrypted_password)


def fetch_public_key(repo):
    """Download RSA public key Travis will use for this repo.

    Travis API docs: http://docs.travis-ci.com/api/#repository-keys
    """
    keyurl = 'https://api.travis-ci.org/repos/{0}/key'.format(repo)
    data = json.loads(urlopen(keyurl).read().decode())
    if 'key' not in data:
        errmsg = "Could not find public key for repo: {}.\n".format(repo)
        errmsg += "Have you already added your GitHub repo to Travis?"
        raise ValueError(errmsg)
    return data['key']


def prepend_line(filepath, line):
    """Rewrite a file adding a line to its beginning."""
    with open(filepath) as f:
        lines = f.readlines()

    lines.insert(0, line)

    with open(filepath, 'w') as f:
        f.writelines(lines)


def load_yaml_config(filepath):
    """Load yaml config file at the given path."""
    with open(filepath) as f:
        return yaml.load(f)


def save_yaml_config(filepath, config):
    """Save yaml config file at the given path."""
    with open(filepath, 'w') as f:
        yaml.dump(config, f, default_flow_style=False)


def update_travis_deploy_password(encrypted_password):
    """Put `encrypted_password` into the deploy section of .travis.yml."""
    config = load_yaml_config(TRAVIS_CONFIG_FILE)

    config['deploy']['password'] = dict(secure=encrypted_password)

    save_yaml_config(TRAVIS_CONFIG_FILE, config)

    line = ('# This file was autogenerated and will overwrite'
            ' each time you run travis_pypi_setup.py\n')
    prepend_line(TRAVIS_CONFIG_FILE, line)


def main(args):
    """Add a PyPI password to .travis.yml so that Travis can deploy to PyPI.

    Fetch the Travis public key for the repo, and encrypt the PyPI password
    with it before adding, so that only Travis can decrypt and use the PyPI
    password.
    """
    public_key = fetch_public_key(args.repo)
    password = args.password or getpass('PyPI password: ')
    update_travis_deploy_password(encrypt(public_key, password.encode()))
    print("Wrote encrypted password to .travis.yml -- you're ready to deploy")


if '__main__' == __name__:
    import argparse
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--repo', default=GITHUB_REPO,
                        help='GitHub repo (default: %s)' % GITHUB_REPO)
    parser.add_argument('--password',
                        help='PyPI password (will prompt if not provided)')

    args = parser.parse_args()
    main(args)


bcc/bcc.py
# -*- coding: utf-8 -*-

"""Main module."""

import numpy as np
import traitlets as tl
import traittypes as tt


class PositiveInt(tl.Int):

    def info(self):
        return u'a positive integer'

    def validate(self, obj, proposal):
        super().validate(obj, proposal)
        if proposal <= 0:
            self.error(obj, proposal)
        return proposal


class PositiveFloat(tl.Float):

    def info(self):
        return u'a positive float'

    def validate(self, obj, proposal):
        super().validate(obj, proposal)
        if proposal <= 0.0:
            self.error(obj, proposal)
        return proposal


class BCC(tl.HasTraits):
    """BCC lattice maps voronoi cells to bin ids
        DOES THE GEOMETRIC WINDOW DEFINE THE GRID, OR
                 1:N contiguous
        option 1: all idx in range(0, len(binner)) valid
        option 2: all points in [lower, upper] have valid indices
        ARE THESE MUTUALLY EXCLUSIVE????
    """
    ndim = PositiveInt()
    sizes = tt.Array(dtype='u8')  # (upper-lower)/width
    nside = tt.Array(dtype='u8')  # 1 + sizes + 1 (for boundary)
    lower = tt.Array(dtype='f8')
    upper = tt.Array(dtype='f8')
    width = tt.Array(dtype='f8')

    def get_bin_index(self, value):
        indices, parity = self._get_bin_indices(value)
        indices += 1
        index = np.sum(self._nside_prod() * indices, axis=-1)
        return np.left_shift(index, 1) + parity

    def get_bin_center(self, index):
        index = np.asarray(index, dtype='i8')
        parity = np.bitwise_and(index, 1)
        indices = np.mod(np.right_shift(index, 1)[..., np.newaxis] //
                         self._nside_prod(), self.nside)
        indices -= 1
        return self._get_bin_center_from_indices(indices, parity)

    def __len__(self):
        return np.prod(self.nside)

    if 1:  # 'HIDDEN METHODS'

        def _get_bin_indices(self, value):
            value = np.asarray(value)
            tmp = value.copy()
            tmp -= self.lower
            tmp /= self.width
            indices = tmp.astype('i8')
            tmp -= indices + 0.5
            corner_indices = indices - (tmp < 0)
            parity = 0.25 * self.ndim < np.sum(np.sign(tmp) * tmp, axis=-1)
            indices = np.where(parity[..., np.newaxis],
                               corner_indices, indices)
            return indices, parity

        def _nside_prod(self):
            _nside_prod = np.ones(self.ndim, dtype='i8')
            _nside_prod[1:] = np.cumprod(self.nside[:-1])
            return _nside_prod

        def _get_bin_center_from_indices(self, indices, parity):
            cen = (self.lower + self.width / 2 + self.width * indices
                   + np.where(parity[..., np.newaxis], self.width / 2.0, 0.0))
            return cen

        @tl.default('ndim')
        def _default_ndim(self):
            if len(self.sizes.shape) is not 1 or self.sizes.shape[0] < 3:
                raise tl.TraitError('sizes must have shape (ndim>=3,)')
            return self.sizes.shape[0]

        @tl.default('sizes')
        def _default_sizes(self):
            raise tl.TraitError('sizes must be specified')

        @tl.default('lower')
        def _default_lower(self):
            return np.zeros((self.ndim,))

        @tl.default('upper')
        def _default_upper(self):
            return np.ones((self.ndim,))

        @tl.default('width')
        def _default_width(self):
            return (self.upper - self.lower) / self.sizes

        @tl.default('nside')
        def _default_nside(self):
            return self.sizes + 2

        @tl.validate('sizes')
        def _validate_sizes(self, proposal):
            return proposal['value']

        @tl.validate('nside')
        def _validate_nside(self, proposal):
            if not np.all(proposal['value'] == self.sizes + 2):
                raise tl.TraitError('bad nside, should be sizes + 2')
            return proposal['value']

        @tl.validate('lower')
        def _validate_lower(self, proposal):
            lower = proposal['value']
            if (self.sizes.shape != lower.shape):
                err = 'sizes and lower must have same shape. '
                err += 'sizes.shape: %s, lower.shape: %s' % (
                    self.sizes.shape, lower.shape)
                raise tl.TraitError(err)
            with self.hold_trait_notifications():
                if np.any(lower >= self.upper):
                    raise tl.TraitError('lower >= upper!')
            return lower

        @tl.validate('upper')
        def _validate_upper(self, proposal):
            upper = proposal['value']
            if (self.sizes.shape != upper.shape):
                err = 'sizes and upper must have same shape. '
                err += 'sizes.shape: %s, upper.shape: %s' % (
                    self.sizes.shape, upper.shape)
                raise tl.TraitError(err)
            with self.hold_trait_notifications():
                if np.any(self.lower >= upper):
                    raise tl.TraitError('lower >= upper!')
            return upper


class _Cubic:
    pass
